spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java

```diff
@@ -75,34 +77,52 @@ public AppDeploymentDeleteServiceInstanceWorkflow(BrokeredServices brokeredServi
-			.mergeWith(collectBoundBackingServices(request))
-			.doOnEach(backingServices -> log.debug("Deleting backing services {} for {}/{}",
-				backingServices, request.getServiceDefinition().getName(), request.getPlan().getName()))
-			.flatMap(backingServicesProvisionService::deleteServiceInstance)
+		return collectBackingServices(request)
+			.collectList()
+			.flatMapMany(backingServices -> {
+				if (!CollectionUtils.isEmpty(backingServices)) {
+					return backingServicesProvisionService.deleteServiceInstance(backingServices);
+				}
+				return Flux.empty();
+			})
 			.doOnComplete(() -> log.debug("Finished deleting backing services for {}/{}",
 				request.getServiceDefinition().getName(), request.getPlan().getName()))
 			.doOnError(exception -> log.error(String.format("Error deleting backing services for %s/%s with error '%s'",
 				request.getServiceDefinition().getName(), request.getPlan().getName(), exception.getMessage()),
 				exception));
 	}
 
-	private Flux<List<BackingService>> collectConfiguredBackingServices(DeleteServiceInstanceRequest request) {
+	private Flux<BackingService> collectBackingServices(DeleteServiceInstanceRequest request) {
+		return collectConfiguredBackingServices(request)
+			.concatWith(collectBoundBackingServices(request))
+			.distinct(BackingService::serviceInstanceNameAndSpaceHashCode);
+	}
+
+	private Flux<BackingService> collectConfiguredBackingServices(DeleteServiceInstanceRequest request) {
 		return getBackingServicesForService(request.getServiceDefinition(), request.getPlan())
-			.flatMapMany(backingServices -> getTargetForService(request.getServiceDefinition(), request.getPlan())
+			.flatMap(backingServices -> getTargetForService(request.getServiceDefinition(), request.getPlan())
 				.flatMap(targetSpec -> targetService.addToBackingServices(backingServices, targetSpec,
 					request.getServiceInstanceId()))
-				.defaultIfEmpty(backingServices));
+				.defaultIfEmpty(backingServices))
+			.flatMapMany(Flux::fromIterable);
 	}
 
-	private Flux<List<BackingService>> collectBoundBackingServices(DeleteServiceInstanceRequest request) {
+	private Flux<BackingService> collectBoundBackingServices(DeleteServiceInstanceRequest request) {
 		return backingAppManagementService.getDeployedBackingApplications(request.getServiceInstanceId())
 			.flatMapMany(Flux::fromIterable)
-			.flatMap(backingApplication -> Flux.fromIterable(backingApplication.getServices())
-				.map(servicesSpec -> BackingService.builder()
-					.serviceInstanceName(servicesSpec.getServiceInstanceName())
-					.build())
-				.collectList());
+			.flatMap(backingApplication -> Mono.justOrEmpty(backingApplication.getServices())
+				.flatMapMany(Flux::fromIterable)
+				.flatMap(servicesSpec -> Mono.justOrEmpty(servicesSpec.getServiceInstanceName()))
+				.map(serviceInstanceName -> {
+					BackingService backingService = BackingService.builder()
+						.serviceInstanceName(serviceInstanceName)
+						.build();
```

Comment:
Nit-picking: I would personally prefer moving the `.build()` to the return line and avoid using setter on backing service. It would give me a signal that this whole section is about building backing service. But the way it is right now is totally fine too.

Reply:
makes sense. I'll move that logic to another function to make it read better.

Reply:
reworked instead of moving to another function